### PR TITLE
tlsroute: make TLSRoute structure consistent with HTTPRoute

### LINKED ISSUE #566: TLSRoute SNI match inconsistency
TLS route:
```yaml
spec:
  rules:
  - matches:
    - snis: ["foo.com"]
    forwardTo:
    - serviceName: httpbin
      port: 443
```

HTTP route:
```yaml
spec:
  hostnames: ["foo.com"]
  rules:
  - forwardTo:
    - serviceName: httpbin
      port: 80
```

The inconsistency here feels a bit odd when compared to HTTPRoute.

Would it make sense to make a `hostnames` field in TLSRoute and remove the `snis` field in matches? Then the specs above would be identical (both looking like the "HTTP Route")

### PR DESCRIPTION
**Pull request by hbagdi**
**What type of PR is this?**

/kind cleanup
/kind api-change

**What this PR does / why we need it**:

This PR makes the structure of TLSRoute similar to HTTPRoute:
- SNIs field has been renamed to Hostnames
- The same field has been pulled up to the Spec struct instead of a match

**Which issue(s) this PR fixes**:

Fixes #566

**Does this PR introduce a user-facing change?**:

```release-note
- Breaking change: the structure of TLSRoute is now consistent with HTTPRoute. 
```

Putting a hold here as there is still work to do. I'd like to get consensus on moving this forward before putting more time in here.
/hold

/cc @howardjohn @bowei @jpeach @robscott

### CODE REVIEW

File: apis/v1alpha2/tlsroute_types.go
Comment (jpeach) [inline apis/v1alpha2/tlsroute_types.go:80]: Would you mind reverting the Go field name to match the JSON/YAML field name? Divergence on the names for fields has led me astray many times in the past :)

Comment (hbagdi) [reply apis/v1alpha2/tlsroute_types.go]: Updated the other way around since the discussion has to be around renaming this field to `hostnames` as well.

File: apis/v1alpha2/tlsroute_types.go
Comment (howardjohn) [inline apis/v1alpha2/tlsroute_types.go:63]: Why not? I think right now in the spec the answer would be "Because you should just use TCPRoute instead in that case", but in the future could we not allow routing on other attributes of the TLS handshake?

I don't feel strongly about this *at all* - our implementation only supports routing by SNI today - just curious.

Comment (howardjohn) [reply apis/v1alpha2/tlsroute_types.go:63]: I see this is also copy+pasted so definitely not blocking this PR at all

**Review by howardjohn — APPROVED** [review]

**Comment by robscott** [comment]
@hbagdi Looks like this needs a rebase. I'm not quite sure how I feel about the sni -> hostname rename here, @danehans or @bowei do you have any strong opinions on this?

**Comment by robscott** [comment]
@hbagdi I've thought about this more and agree that this change makes sense. We discussed this briefly at community meeting today and it didn't sound like there was any opposition to this change, so I'm happy to LGTM once it's rebased.

**Comment by hbagdi** [comment]
@jpeach @robscott I've rebased and updated the release notes. PTAL.

**Comment by robscott** [comment]
Thanks!

/lgtm
/approve

**Pull request merged**
